File: packages/core/src/random.ts
/**
 * Lightweight random utilities — replaces @faker-js/faker (~5KB vs 5MB).
 * Works in both browser and Node.js / Cloudflare Workers.
 */

// ─── Mini word dictionary (~200 words) ──────────────────

const WORDS = [
    'alpha', 'bravo', 'charlie', 'delta', 'echo', 'foxtrot', 'golf', 'hotel',
    'india', 'juliet', 'kilo', 'lima', 'mike', 'november', 'oscar', 'papa',
    'quebec', 'romeo', 'sierra', 'tango', 'uniform', 'victor', 'whiskey',
    'xray', 'yankee', 'zulu', 'apple', 'banana', 'cherry', 'dragon',
    'eagle', 'falcon', 'grape', 'hawk', 'iron', 'jade', 'knight', 'lemon',
    'mango', 'night', 'ocean', 'pearl', 'quartz', 'river', 'storm', 'tiger',
    'umbra', 'venom', 'whale', 'xenon', 'yeti', 'zebra', 'amber', 'blaze',
    'coral', 'dusk', 'ember', 'frost', 'glow', 'haze', 'ivory', 'jet',
    'karma', 'lotus', 'mist', 'neon', 'onyx', 'prism', 'quest', 'rune',
    'silk', 'torch', 'ultra', 'vivid', 'wind', 'pixel', 'solar', 'lunar',
    'cyber', 'nexus', 'pulse', 'spark', 'steel', 'stone', 'swift', 'flux',
    'blitz', 'crypt', 'drift', 'forge', 'grain', 'haven', 'index', 'joule',
    'knot', 'latch', 'mesh', 'node', 'orbit', 'phase', 'query', 'relay',
    'scope', 'trace', 'unity', 'valve', 'warp', 'yield', 'zinc', 'axis',
    'bolt', 'cage', 'data', 'edge', 'fiber', 'grid', 'hash', 'input',
    'joint', 'key', 'link', 'mode', 'null', 'output', 'path', 'queue',
    'root', 'seed', 'token', 'unit', 'void', 'wire', 'zero', 'bit',
    'byte', 'cell', 'core', 'disk', 'font', 'gate', 'heap', 'icon',
    'jack', 'kern', 'load', 'mask', 'nano', 'page', 'rack', 'scan',
    'tape', 'user', 'view', 'wrap', 'zone', 'arch', 'band', 'chip',
    'dock', 'exit', 'flag', 'gear', 'hook', 'iris', 'jolt', 'kite',
    'lamp', 'maze', 'nest', 'opal', 'pine', 'ring', 'slab', 'tile',
    'undo', 'veil', 'wave', 'apex', 'barn', 'cape', 'dome', 'echo',
    'fern', 'glen', 'hill', 'isle', 'jump', 'knob', 'leaf', 'mill',
    'nova', 'oval', 'pond', 'quay', 'reef', 'sand', 'tent', 'urge',
];

const FIRST_NAMES = [
    'James', 'Mary', 'John', 'Patricia', 'Robert', 'Jennifer', 'Michael',
    'Linda', 'David', 'Elizabeth', 'William', 'Barbara', 'Richard', 'Susan',
    'Joseph', 'Jessica', 'Thomas', 'Sarah', 'Charles', 'Karen',
];

const LAST_NAMES = [
    'Smith', 'Johnson', 'Williams', 'Brown', 'Jones', 'Garcia', 'Miller',
    'Davis', 'Rodriguez', 'Martinez', 'Hernandez', 'Lopez', 'Gonzalez',
    'Wilson', 'Anderson', 'Thomas', 'Taylor', 'Moore', 'Jackson', 'Martin',
];

const DOMAINS = [
    'test.com', 'example.org', 'demo.net', 'sample.io', 'mock.dev',
    'fake.co', 'stub.app', 'local.test',
];

// ─── Core random functions ──────────────────────────────

/** Pick a random element from array */
export function pick<T>(arr: readonly T[]): T {
    return arr[Math.floor(Math.random() * arr.length)];
}

/** Generate a UUID v4 */
export function uuid(): string {
    if (typeof crypto !== 'undefined' && crypto.randomUUID) {
        return crypto.randomUUID();
    }
    // Fallback
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
        const r = (Math.random() * 16) | 0;
        const v = c === 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}

/** Random word from built-in dictionary */
export function word(): string {
    return pick(WORDS);
}

/** Multiple random words joined by space */
export function words(n: number): string {
    return Array.from({ length: n }, () => word()).join(' ');
}

/** Random sentence (5-12 words, capitalized, with period) */
export function sentence(): string {
    const n = int(5, 12);
    const s = words(n);
    return s.charAt(0).toUpperCase() + s.slice(1) + '.';
}

/** Random integer in [min, max] inclusive */
export function int(min: number, max: number): number {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

/** Random float in [min, max) */
export function float(min: number, max: number): number {
    return Math.random() * (max - min) + min;
}

/** Random boolean */
export function bool(): boolean {
    return Math.random() < 0.5;
}

/** Random date between from and to */
export function date(
    from: Date = new Date('2020-01-01'),
    to: Date = new Date(),
): Date {
    const fromMs = from.getTime();
    const toMs = to.getTime();
    return new Date(fromMs + Math.random() * (toMs - fromMs));
}

/** Random email */
export function email(): string {
    const first = pick(FIRST_NAMES).toLowerCase();
    const last = pick(LAST_NAMES).toLowerCase();
    const domain = pick(DOMAINS);
    return `${first}.${last}${int(1, 999)}@${domain}`;
}

/** Random IPv4 address */
export function ipv4(): string {
    return `${int(1, 254)}.${int(0, 255)}.${int(0, 255)}.${int(1, 254)}`;
}

/** Random URI */
export function uri(): string {
    return `https://${word()}.example.com/${word()}/${word()}`;
}

/** Random string of given length (alphanumeric) */
export function randomString(length: number): string {
    const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
        result += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return result;
}

/** Random full name */
export function fullName(): string {
    return `${pick(FIRST_NAMES)} ${pick(LAST_NAMES)}`;
}

/** Random phone number */
export function phone(): string {
    return `+1${int(200, 999)}${int(100, 999)}${int(1000, 9999)}`;
}

File: packages/container/internal/generator/random.go
package generator

import (
	"fmt"
	"math/rand/v2"
	"strings"
	"time"

	"github.com/google/uuid"
)

// ─── Mini word dictionary (~200 words) ──────────────────

var words = []string{
	"alpha", "bravo", "charlie", "delta", "echo", "foxtrot", "golf", "hotel",
	"india", "juliet", "kilo", "lima", "mike", "november", "oscar", "papa",
	"quebec", "romeo", "sierra", "tango", "uniform", "victor", "whiskey",
	"xray", "yankee", "zulu", "apple", "banana", "cherry", "dragon",
	"eagle", "falcon", "grape", "hawk", "iron", "jade", "knight", "lemon",
	"mango", "night", "ocean", "pearl", "quartz", "river", "storm", "tiger",
	"umbra", "venom", "whale", "xenon", "yeti", "zebra", "amber", "blaze",
	"coral", "dusk", "ember", "frost", "glow", "haze", "ivory", "jet",
	"karma", "lotus", "mist", "neon", "onyx", "prism", "quest", "rune",
	"silk", "torch", "ultra", "vivid", "wind", "pixel", "solar", "lunar",
	"cyber", "nexus", "pulse", "spark", "steel", "stone", "swift", "flux",
	"blitz", "crypt", "drift", "forge", "grain", "haven", "index", "joule",
	"knot", "latch", "mesh", "node", "orbit", "phase", "query", "relay",
	"scope", "trace", "unity", "valve", "warp", "yield", "zinc", "axis",
}

var firstNames = []string{
	"James", "Mary", "John", "Patricia", "Robert", "Jennifer", "Michael",
	"Linda", "David", "Elizabeth", "William", "Barbara", "Richard", "Susan",
	"Joseph", "Jessica", "Thomas", "Sarah", "Charles", "Karen",
}

var lastNames = []string{
	"Smith", "Johnson", "Williams", "Brown", "Jones", "Garcia", "Miller",
	"Davis", "Rodriguez", "Martinez", "Hernandez", "Lopez", "Gonzalez",
	"Wilson", "Anderson", "Thomas", "Taylor", "Moore", "Jackson", "Martin",
}

var domains = []string{
	"test.com", "example.org", "demo.net", "sample.io", "mock.dev",
	"fake.co", "stub.app", "local.test",
}

const alphanumeric = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

// ─── Core random functions ──────────────────────────────

// Pick returns a random element from a slice.
func Pick[T any](arr []T) T {
	return arr[rand.IntN(len(arr))]
}

// UUID generates a UUID v4.
func UUID() string {
	return uuid.New().String()
}

// Word returns a random word from the built-in dictionary.
func Word() string {
	return Pick(words)
}

// Words returns n random words joined by space.
func Words(n int) string {
	var b strings.Builder
	for i := 0; i < n; i++ {
		if i > 0 {
			b.WriteByte(' ')
		}
		b.WriteString(Word())
	}
	return b.String()
}

// IntRange returns a random integer in [min, max] inclusive.
func IntRange(min, max int) int {
	return min + rand.IntN(max-min+1)
}

// FloatRange returns a random float in [min, max).
func FloatRange(min, max float64) float64 {
	return min + rand.Float64()*(max-min)
}

// RandomDate returns a random date between 2020-01-01 and now.
func RandomDate() time.Time {
	from := time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC)
	to := time.Now().UTC()
	diff := to.Sub(from)
	return from.Add(time.Duration(rand.Int64N(int64(diff))))
}

// Email generates a random email address.
func Email() string {
	first := strings.ToLower(Pick(firstNames))
	last := strings.ToLower(Pick(lastNames))
	domain := Pick(domains)
	return fmt.Sprintf("%s.%s%d@%s", first, last, IntRange(1, 999), domain)
}

// IPv4 generates a random IPv4 address.
func IPv4() string {
	return fmt.Sprintf("%d.%d.%d.%d", IntRange(1, 254), IntRange(0, 255), IntRange(0, 255), IntRange(1, 254))
}

// URI generates a random URI.
func URI() string {
	return fmt.Sprintf("https://%s.example.com/%s/%s", Word(), Word(), Word())
}

// RandomString generates a random alphanumeric string of given length.
func RandomString(length int) string {
	b := make([]byte, length)
	for i := range b {
		b[i] = alphanumeric[rand.IntN(len(alphanumeric))]
	}
	return string(b)
}

// FullName generates a random full name.
func FullName() string {
	return fmt.Sprintf("%s %s", Pick(firstNames), Pick(lastNames))
}

// HashStr computes a fast djb2 hash of a string — used for payload dedup.
func HashStr(s string) uint32 {
	var h uint32 = 5381
	for i := 0; i < len(s); i++ {
		h = (h << 5) + h + uint32(s[i])
	}
	return h
}
